File: Python/day15/AOC.py
def parse_input(input_data):
    grid, moves = input_data.split("\n\n")
    grid = [list(line) for line in grid.splitlines()]
    moves = moves.replace("\n", "")
    return grid, moves

def find_robot_position(grid):
    for r, row in enumerate(grid):
        for c, cell in enumerate(row):
            if cell == "@":
                return (r, c)
            

def calculate_gps(grid):
    gps_sum = 0
    for r, row in enumerate(grid):
        for c, cell in enumerate(row):
            if cell in "O[":
                gps_sum += 100 * r + c
    return gps_sum

def part1(input):
    grid, moves = parse_input(input)
    r, c = find_robot_position(grid)
    directions = {"^": (-1, 0), "v": (1, 0), "<": (0, -1), ">": (0, 1)}
    for move in moves:
        dr, dc = directions[move]
        cr = r
        cc = c
        boxes = []
        moveable = True
        while True:
            cr += dr
            cc += dc
            char = grid[cr][cc]
            if char == "#":
                moveable = False
                break
            if char == "O":
                boxes.append((cr, cc))
            if char == ".":
                break
        if not moveable:
            continue
        grid[r][c] = "."
        grid[r + dr][c + dc] = "@"
        for br, bc in boxes:
            grid[br + dr][bc + dc] = "O"
        r += dr
        c += dc
    return calculate_gps(grid)

def scale_up(input):
    expand = {".": "..", "#": "##", "O": "[]", "@": "@."}
    grid = []
    for row in input:
        new_row = ""
        for c in row:
            new_row += expand[c]
        grid.append(list(new_row))
    return grid

def part2(input):
    grid, moves = parse_input(input)
    grid = scale_up(grid)
    r, c = find_robot_position(grid)
    directions = {"^": (-1, 0), "v": (1, 0), "<": (0, -1), ">": (0, 1)}
    for move in moves:
        dr, dc = directions[move]
        boxes = [(r, c)]
        moveable = True
        for cr, cc in boxes:
            nr = cr + dr
            nc = cc + dc
            if (nr, nc) in boxes:
                continue
            char = grid[nr][nc]
            if char == "#":
                moveable = False
                break
            if char == "[":
                boxes.append((nr, nc))
                boxes.append((nr, nc + 1))
            if char == "]":
                boxes.append((nr, nc - 1))
                boxes.append((nr, nc))
        if not moveable:
            continue
        copy = [row[:] for row in grid]
        grid[r][c] = "."
        grid[r + dr][c + dc] = "@"
        for br, bc in boxes[1:]:
            grid[br][bc] = "."
        for br, bc in boxes[1:]:
            grid[br + dr][bc + dc] = copy[br][bc]
        r += dr
        c += dc
    return calculate_gps(grid)

def main():
    input = open("input.txt", "r").read()
    part1_ans = part1(input)
    part2_ans = part2(input)
    print(f"Part 1: {part1_ans}")
    print(f"Part 2: {part2_ans}")

if __name__ == "__main__":
    main()

File: Go/day15/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Point struct {
	r, c int
}

type Points []Point

func (p Points) contains(point Point) bool {
	for _, p := range p {
		if p == point {
			return true
		}
	}
	return false
}

type Grid map[Point]rune

func parseInput(input string) (Grid, []rune, Point) {
	grid := make(Grid, 0)
	var start Point
	var moves []rune
	parts := strings.Split(input, "\n\n")
	for r, line := range strings.Split(parts[0], "\n") {
		for c, char := range line {
			grid[Point{r, c}] = char
			if char == '@' {
				start = Point{r, c}
			}
		}
	}
	moves = []rune(strings.ReplaceAll(parts[1], "\n", ""))
	return grid, moves, start
}

func (g Grid) calculateGPS() int {
	sum := 0
	for point, char := range g {
		if char == 'O' || char == '[' {
			sum += 100*point.r + point.c
		}
	}
	return sum
}

func part1(input string) int {
	grid, moves, start := parseInput(input)
	directions := map[rune]Point{
		'^': {-1, 0},
		'v': {1, 0},
		'<': {0, -1},
		'>': {0, 1},
	}

	for _, move := range moves {
		boxes := Points{}
		dr, dc := directions[move].r, directions[move].c
		cr, cc := start.r, start.c
		moveable := true
		for true {
			cr += dr
			cc += dc
			char := grid[Point{cr, cc}]
			if char == '#' {
				moveable = false
				break
			}
			if char == '.' {
				break
			}
			if char == 'O' {
				boxes = append(boxes, Point{cr, cc})
			}
		}
		if !moveable {
			continue
		}
		grid[start] = '.'
		grid[Point{start.r + dr, start.c + dc}] = '@'
		for _, box := range boxes {
			grid[Point{box.r + dr, box.c + dc}] = 'O'
		}
		start = Point{start.r + dr, start.c + dc}
	}

	return grid.calculateGPS()
}

func expandGrid(input string) string {
	sb := strings.Builder{}
	expansion := map[rune]string{'.': "..", '#': "##", 'O': "[]", '@': "@."}
	for _, line := range strings.Split(input, "\n") {
		for _, c := range line {
			sb.WriteString(expansion[c])
		}
		sb.WriteString("\n")
	}
	return sb.String()
}

func part2(input string) int {
	parts := strings.Split(input, "\n\n")
	expanded := expandGrid(parts[0])
	grid, moves, start := parseInput(expanded + "\n\n" + parts[1])
	directions := map[rune]Point{
		'^': {-1, 0},
		'v': {1, 0},
		'<': {0, -1},
		'>': {0, 1},
	}
	for _, move := range moves {
		boxes := Points{start}
		dr, dc := directions[move].r, directions[move].c
		moveable := true
		for i := 0; i < len(boxes); i++ {
			nr := boxes[i].r + dr
			nc := boxes[i].c + dc
			if boxes.contains(Point{nr, nc}) {
				continue
			}
			char := grid[Point{nr, nc}]
			if char == '#' {
				moveable = false
				break
			}
			if char == '[' {
				boxes = append(boxes, Point{nr, nc})
				boxes = append(boxes, Point{nr, nc + 1})
			}
			if char == ']' {
				boxes = append(boxes, Point{nr, nc})
				boxes = append(boxes, Point{nr, nc - 1})
			}
		}
		if !moveable {
			continue
		}
		chars := map[Point]rune{}
		grid[start] = '.'
		for _, box := range boxes[1:] {
			chars[box] = grid[box]
			grid[box] = '.'
		}

		for _, box := range boxes[1:] {
			grid[Point{box.r + dr, box.c + dc}] = chars[box]
		}
		grid[Point{start.r + dr, start.c + dc}] = '@'
		start = Point{start.r + dr, start.c + dc}
	}

	return grid.calculateGPS()

}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := string(input)
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
